### PR TITLE
Add Retry Logic To Zuul Using Spring Retry

### PR DESCRIPTION
**Pull request by ryanjbaxter**
This PR adds retry logic back to Zuul when using the Apache HTTP Client or the Ok HTTP Client.  This is the last piece to fixing #1295.

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java [hunk 1]
```diff
@@ -66,21 +78,36 @@ protected OkHttpClient createDelegate(IClientConfig config) {
...
+		return this.executeWithRetry(ribbonRequest, new RetryCallback() {
+			@Override
+			public OkHttpRibbonResponse doWithRetry(RetryContext context) throws Exception {
+				//on retries the policy will choose the server and set it in the context
+				//extract the server and update the request being made
+				OkHttpRibbonRequest newRequest = ribbonRequest;
+				if(context instanceof LoadBalancedRetryContext) {
+					ServiceInstance service = ((LoadBalancedRetryContext)context).getServiceInstance();
+					if(service != null) {
+						//Reconstruct the request URI using the host and port set in the retry context
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java]: So this retries without trying a different server?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/support/RetryableLoadBalancingClient.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * A load balancing client which uses Spring Retry to retry failed requests.
+ * @author Ryan Baxter
+ */
+public abstract class RetryableLoadBalancingClient<S extends ContextAwareRequest, T extends IResponse, D>
+		extends AbstractLoadBalancingClient<S, T, D> implements ServiceInstanceChooser {
+
+	protected LoadBalancedRetryPolicyFactory loadBalancedRetryPolicyFactory =
+			new LoadBalancedRetryPolicyFactory.NeverRetryFactory();
+
+	@Deprecated
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/support/RetryableLoadBalancingClient.java]: Why introduce deprecated constructors?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java (hunk 1 above)
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java]: You have tests later for next server. Can you point me to where next server is determined?

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/apache/RibbonLoadBalancingHttpClientTests.java [hunk 3]
```diff
@@ -129,6 +163,169 @@ public void testUpdatedTimeouts()
 		assertThat(result.getSocketTimeout(), is (50));
 	}
 
+	@Test
+	public void testNeverRetry() throws Exception {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/apache/RibbonLoadBalancingHttpClientTests.java]: Awesome tests!

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java (hunk 1 above)
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/okhttp/OkHttpLoadBalancingClient.java]: It happens here https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonLoadBalancedRetryPolicyFactory.java#L73

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/support/RetryableLoadBalancingClient.java (hunk 2 above)
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/support/RetryableLoadBalancingClient.java]: I guess we can remove the `@Deprecated` annotations.  The constructors in the subclasses (which I don't want to remove in Camden) need to call `super`.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/support/RetryableLoadBalancingClient.java]: NP, that's valid.

**Pull request merged**